modules/linear_boltzmann_solvers/executors/pi_keigen.cc

```diff
@@ -182,5 +191,78 @@ PowerIterationKEigen::SetLBSScatterSource(const std::vector<double>& input,
     front_gs_, q_moments_local_, input, lbs_solver_.DensitiesLocal(), source_flags);
 }
 
+void
+PowerIterationKEigen::WriteRestartData(bool force)
+{
+  // Check if restarts are enabled
+  auto write_restart_time_interval = lbs_solver_.Options().write_restart_time_interval;
+  if (write_restart_time_interval == 0)
+    return;
+
+  // Check if enough time has elapsed since the last restart write
+  auto& last_restart_write_time = lbs_solver_.GetLastRestartWriteTime();
+  auto now = std::chrono::system_clock::now().time_since_epoch();
+  size_t now_secs = std::chrono::duration_cast<std::chrono::seconds>(now).count();
+  if (((now_secs - last_restart_write_time) < write_restart_time_interval) and not force)
+    return;
+
+  std::string fbase = lbs_solver_.Options().write_restart_directory_name + "/" +
+                      lbs_solver_.Options().write_restart_file_stem;
+  std::string fname = fbase + std::to_string(opensn::mpi_comm.rank()) + ".r";
```

Comment:
I originally implemented something similar to this, but ultimately decided that having the individual pieces of the path available made for better error messages.

Reply:
Could you point me to one of those error messages? If we use something like `std::filesystem::path` it should be trivial to get the info you need for any error message.

Reply:
That's a very good point. Let me see about fixing this.

Reply:
I've removed most of the restart parameters for files and paths in favor of `read_restart_path` and `write_restart_path` similar to suggestions by @zhardy-lanl below. I didn't create a `GetRestartDumpPath()` method so that individual solvers can tailor the path for their needs, but it's still a much cleaner interface.